enterprise/tailnet/pgcoord.go

```diff
@@ -0,0 +1,1208 @@
+		bk := bKey{client: m.client, agent: m.agent}
+		bestM, ok := best[bk]
+		if !ok || m.updatedAt.After(bestM.updatedAt) {
+			best[bk] = m
+		}
+	}
+	nodes := make([]*agpl.Node, 0, len(best))
+	for _, m := range best {
+		nodes = append(nodes, m.node)
+	}
+	return nodes
+}
+
+// querier is responsible for monitoring pubsub notifications and querying the database for the mappings that all
+// connected clients and agents need.  It also checks heartbeats and withdraws mappings from coordinators that have
+// failed heartbeats.
+type querier struct {
+	ctx            context.Context
+	logger         slog.Logger
+	pubsub         pubsub.Pubsub
+	store          database.Store
+	newConnections chan *connIO
+
+	workQ      *workQ[mKey]
+	heartbeats *heartbeats
+	updates    <-chan struct{}
+
+	mu      sync.Mutex
+	mappers map[mKey]*countedMapper
+}
+
+type countedMapper struct {
+	*mapper
+	count  int
+	cancel context.CancelFunc
+}
+
+func newQuerier(
+	ctx context.Context, logger slog.Logger,
+	ps pubsub.Pubsub, store database.Store,
+	self uuid.UUID, newConnections chan *connIO, numWorkers int,
+	firstHeartbeat chan<- struct{},
+) *querier {
+	updates := make(chan struct{})
+	q := &querier{
+		ctx:            ctx,
+		logger:         logger.Named("querier"),
+		pubsub:         ps,
+		store:          store,
+		newConnections: newConnections,
+		workQ:          newWorkQ[mKey](ctx),
+		heartbeats:     newHeartbeats(ctx, logger, ps, store, self, updates, firstHeartbeat),
+		mappers:        make(map[mKey]*countedMapper),
+		updates:        updates,
+	}
+	go q.subscribe()
+	go q.handleConnIO()
+	for i := 0; i < numWorkers; i++ {
+		go q.worker()
+	}
+	go q.handleUpdates()
+	return q
+}
+
+func (q *querier) handleConnIO() {
+	for {
+		select {
+		case <-q.ctx.Done():
+			return
+		case c := <-q.newConnections:
+			q.newConn(c)
+		}
+	}
+}
+
+func (q *querier) newConn(c *connIO) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	mk := mKey{
+		agent: c.agent,
+		// if client is Nil, this is an agent connection, and it wants the mappings for all the clients of itself
+		clientsOfAgent: c.client == uuid.Nil,
+	}
+	cm, ok := q.mappers[mk]
+	if !ok {
+		ctx, cancel := context.WithCancel(q.ctx)
+		mpr := newMapper(ctx, q.logger, mk, q.heartbeats)
+		cm = &countedMapper{
+			mapper: mpr,
+			count:  0,
+			cancel: cancel,
+		}
+		q.mappers[mk] = cm
+		// we don't have any mapping state for this key yet
+		q.workQ.enqueue(mk)
+	}
+	if err := sendCtx(cm.ctx, cm.add, c); err != nil {
+		return
+	}
+	cm.count++
+	go q.waitForConn(c)
+}
+
+func (q *querier) waitForConn(c *connIO) {
+	<-c.ctx.Done()
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	mk := mKey{
+		agent: c.agent,
+		// if client is Nil, this is an agent connection, and it wants the mappings for all the clients of itself
+		clientsOfAgent: c.client == uuid.Nil,
+	}
+	cm := q.mappers[mk]
+	if err := sendCtx(cm.ctx, cm.del, c); err != nil {
+		return
+	}
+	cm.count--
+	if cm.count == 0 {
+		cm.cancel()
+		delete(q.mappers, mk)
+	}
+}
```

Comment:
The locking logic here between `newConn` and `waitForConn` is tricky to follow. Is there a particular reason we lock, spawn a new goroutine, and unlock, only to immediately lock and unlock again?

Reply:
maybe a rename would make this more clear --- waitForConn doesn't do anything immediately, it waits for the connIO's context to expire, and then cleans it up.  It's on a separate goroutine so that we don't block newConn from returning.

How do you feed about `cleanupConn()`?

Reply:
Ah, the key here is `<-c.ctx.Done()` -- that's easy to skip over.
`cleanupConn` does make more sense in this context.